Update Pagination component (#954)
Workday/canvas-kit: modules/_labs/pagination/react/README.md

  <img src="https://img.shields.io/badge/LABS-beta-orange" alt="LABS: Beta" />
</a>  This component is work in progress and currently in pre-release.

> `Pagination` is a composable component for handling navigation between pages in a range.

Author: Yeah, this was written before we were using the term "compound component." I'll update it. Good catch. 👍🏻

## Usage

`Pagination` is a composable component, meaning it can be put together and used in a variety of
different ways. We'll first cover basic usage examples followed by advanced usage. Each of the
examples is intended to be standalone, so you can skip to the section that best fits your use case.

These examples are designed to provide copy-and-paste snippets to help you get up and running
quickly without much explanation. Further detail on individual component props, behavior, and
accessibility can be found in the [Components](#components) section.

### Basic Usage

Below are examples we expect will work for most use cases. Please refer to the Advanced Usage
section for details on supporting custom implementations.

#### Pagination with Step Controls

In this example, the component will only display the step controls (next page and previous page) and
the page list buttons.

```tsx
import * as React from 'react';

> import Pagination from '@workday/canvas-kit-labs-react-pagination';

Reviewer: The Pagination imports will need to be updated to `import {Pagination}` if there's no longer a default export.

Author: Great catch!

feel free to explore this section. While these examples are not exhaustive, they provide additional
insight into what's possible.

#### Controllable State Using Pagination Model

If you want the `Pagination` component to handle its state and actions internally and hook into page
change events, the basic usage examples above should be sufficient. However, if you need to control
the state externally, or if you want more control over the model, you'll need to use a controllable
state pattern by using the `Pagination` model. You can create the model with the
`usePaginationModel` hook provided:

```tsx
const model = usePaginationModel({
  lastPage={100}
  initialCurrentPage={6}
  rangeSize={3}
  onPageChange={pageNumber => console.log(pageNumber)}
});
```

Notice that the hook takes the same props as the `Pagination` component (with the exception of
`aria-label` which isn't needed to build the model). This allows you to quickly convert your
implementation to this pattern if needed without a major refactor. You can then supply this model to
the `Pagination` component.

```tsx
const ControllablePagination = () => {
  // create the model

>   const model = usePaginationModel({
>   lastPage={100}
>   initialCurrentPage={6}
>   rangeSize={3}
>   onPageChange={pageNumber => console.log(pageNumber)}
> });

Reviewer 2: Spacing is off here:

```suggestion
  const model = usePaginationModel({
    lastPage={100}
    initialCurrentPage={6}
    rangeSize={3}
    onPageChange={pageNumber => console.log(pageNumber)}
  });
```

  return (
    <Pagination aria-label="Pagination" model={model}>
      {/* child components go here */}
    </Pagination>
  );
}
```

Notice that when providing a model, you no longer need to provide the other props (other than
`aria-label`). Let's look at a complete example where you might want to use this controllable
pattern.

First, we'll create a table that will fetch some mock data:

```tsx

> ```

Reviewer 2: > This example intentionally left blank

Author: It's...uh... the essay portion of the exam. Once they've read the docs, they should be able to fill in their own example here, and I'll grade it. 😉

## Components

### Pagination

The `Pagination` component is a wrapper for a React context provider and `nav` element. Child
components such as `Pagination.StepToPreviousButton`, `Pagination.PageList`, etc. subscribe to that

> context.

Reviewer 2: Do we want to talk about this as the `model`? In this doc, this is referred to as "`Pagination`'s context` vs "`PaginationModel` context". The context is the reference to the shared model. The language used here is simpler. I've been trying to carefully use the same language for these things to avoid confusion (i.e. What's the difference between the model and the context?)

Author: Probably not. I wrote this before your doc was created. I'll link out to it. Good call. 👍🏻

#### Usage

```tsx
<Pagination
  aria-label="Pagination"
  lastPage={100}
  initialCurrentPage={6}
  rangeSize={3}
  onPageChange={pageNumber => console.log(pageNumber)}
>
  {/* child components go here */}
</Pagination>
```

#### Component Props

| name               | type                           | required?  | default | recommended                              |
| ------------------ | ------------------------------ | ---------- | ------- | ---------------------------------------- |
| aria-label         | `string`                       | ✅ `true`  | n/a     | "Pagination" (and translated equivalent) |
| lastPage           | `number`                       | ✅ `true`  | n/a     | n/a                                      |
| initialCurrentPage | `number`                       | 🚫 `false` | 1       | n/a                                      |
| rangeSize          | `number`                       | 🚫 `false` | 5       | n/a                                      |
| firstPage          | `number`                       | 🚫 `false` | 1       | n/a                                      |

> | onPageChange       | `(pageNumber: number) => void` | 🚫 `false` | n/a     | n/a                                      |

Author: It needs to be added. The API changed. 👍🏻

| ---------- | -------- | --------- | ------- | ---------------------------------- |
| aria-label | `string` | ✅ `true` | n/a     | "Last" (and translated equivalent) |

This component also supports all `IconButton` props.

### Pagination.PageList

`Pagination.PageList` is an ordered list (`ol`) that subscribes to `Pagination`'s context. This
allows it generate the `range` of page numbers. It also handles spacing between the child elements.
This component will accept either child elements or functional children (sometimes called the render
prop pattern). It's likely you'll want to use the functional children, but in the off-chance you
need a static list of items, this component will support it. The usage section below will provide
examples of both.

#### Usage

The functional children snippet below will likely be the most common use case.

```tsx
<Pagination.PageList>
  {({state}) =>
    state.range.map(pageNumber => (
      <Pagination.PageListItem key={pageNumber}>
        <Pagination.PageButton aria-label={`Page ${pageNumber}`} pageNumber={pageNumber} />
      </Pagination.PageListItem>
    ))
  }
</Pagination.PageList>
```

#### Component Props

> | name     | type                                                              | required?  | default | recommended |
> | -------- | ----------------------------------------------------------------- | ---------- | ------- | ----------- |
> | children | `(model: PaginationModel) => React.ReactNode[] | React.ReactNode` | 🚫 `false` | n/a     | n/a         |

Reviewer 2: ```suggestion
| name     | type                                                              | required?  | default | recommended |
| -------- | ----------------------------------------------------------------- | ---------- | ------- | ----------- |
| children | `(model: PaginationModel) => React.ReactNode[] \| React.ReactNode` | 🚫 `false` | n/a     | n/a         |
```
renders:
| name     | type                                                              | required?  | default | recommended |
| -------- | ----------------------------------------------------------------- | ---------- | ------- | ----------- |
| children | `(model: PaginationModel) => React.ReactNode[] \| React.ReactNode` | 🚫 `false` | n/a     | n/a         |

vs:
| name     | type                                                              | required?  | default | recommended |
| -------- | ----------------------------------------------------------------- | ---------- | ------- | ----------- |
| children | `(model: PaginationModel) => React.ReactNode[] | React.ReactNode` | 🚫 `false` | n/a     | n/a         |

#### Usage

```tsx
<Pagination.GoToForm>{/* child elements go here */}</Pagination.GoToForm>
```

#### Component Props

This component supports
[all native HTMLFormElement props](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#Attributes).

### Pagination.GoToTextInput

#### Usage

```tsx
<Pagination.GoToTextInput />
```

#### Component Props

This component supports all `TextInput` props.

### Pagination.GoToLabel

`Pagination.GoToLabel` is a styled `label` element that subscribes to the `Pagination` context. This
allows it to pass the `Pagination` context to child elements.

#### Usage

This component will accept either child elements or functional children (sometimes called the render
prop pattern). It's likely you'll want to use the functional children, but in the off-chance you
need static child elements, this component will support it. Examples of both patterns are provided
below.

**Functional Children**

Use this pattern when you need access to the state in the `Pagination` context for your text.

```tsx
<Pagination.GoToLabel>{({state}) => `of ${state.lastPage} results`}</Pagination.GoToLabel>
```

#### Component Props

> | name     | type                                                            | required?  | default | recommended |
> | -------- | --------------------------------------------------------------- | ---------- | ------- | ----------- |
> | htmlFor  | `string`                                                        | 🚫 `false` | n/a     | n/a         |
> | children | `(model: PaginationModel) => React.ReactNode | React.ReactNode` | 🚫 `false` | n/a     | n/a         |

Reviewer 2: ```suggestion
| name     | type                                                            | required?  | default | recommended |
| -------- | --------------------------------------------------------------- | ---------- | ------- | ----------- |
| htmlFor  | `string`                                                        | 🚫 `false` | n/a     | n/a         |
| children | `(model: PaginationModel) => React.ReactNode \| React.ReactNode` | 🚫 `false` | n/a     | n/a         |
```

This component also supports
[all native HTMLLabelElement props](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label#Attributes).

### Pagination.AdditionalDetails

`Pagination.AdditionalDetails` is a styled `div` container that subscribes to the `Pagination`
context. This allows it to pass the `Pagination` context to child elements. It is also an
`aria-live` region that announces the current page update to screen readers. Because of that, it's
important to always include it in your `Pagination` component.

#### Usage

This component will accept either child elements or functional children (sometimes called the render
prop pattern). It's likely you'll want to use the functional children, but in the off-chance you
need static child elements, this component will support it.

```tsx
<Pagination.AdditionalDetails>
  {({state}) => `${state.rangeMin}-${state.rangeMax} of ${state.lastPage} results`}
</Pagination.AdditionalDetails>
```

#### Component Props

> | name              | type                                                            | required?  | default | recommended |
> | ----------------- | --------------------------------------------------------------- | ---------- | ------- | ----------- |
> | shouldHideDetails | `boolean`                                                       | 🚫 `false` | n/a     | n/a         |
> | children          | `(model: PaginationModel) => React.ReactNode | React.ReactNode` | ✅ `true`  | n/a     | n/a         |

Reviewer 2: ```suggestion
| name              | type                                                            | required?  | default | recommended |
| ----------------- | --------------------------------------------------------------- | ---------- | ------- | ----------- |
| shouldHideDetails | `boolean`                                                       | 🚫 `false` | n/a     | n/a         |
| children          | `(model: PaginationModel) => React.ReactNode \| React.ReactNode` | ✅ `true`  | n/a     | n/a         |
```

    jumpToLast,
    stepToNext,
    stepToPrevious,
    setCurrentPage,
    goToPage,
  };
};

```

#### State

The `state` describes the current state of the `Pagination` component. As events are fired, the
state is updated, and the component will re-render. Below are the `state` keys and descriptions of
their values:

- `firstPage` - the page number for the first page
- `currentPage` - the page number for the current page
- `lastPage` - the page number for the last page (it can also be used as a total page count)
- `range` - an array of page numbers included in the visible range
- `rangeSize` - the size of the visible range
- `rangeMin` - the first page number in the visible range
- `rangeMax` - the last page number in the visible range

#### Events

The Model's `events` describe behaviors that act on `state`. Below are the `events` keys and
descriptions of their values:

- `jumpToFirst` - sets the current page to the first page
- `jumpToLast` - sets the current page to the last page
- `stepToNext` - increments the current page by 1
- `stepToPrevious` - decrements the current page by 1
- `setCurrentPage` - sets the current page to a given page number (no safeguards)

> - `goToPage` - sets the current page to a given page number\* (with safeguards)

Reviewer 2: There's some inconsistency with these event names: `jumpTo*`, `stepTo*`, `goTo*`. Tabs is using the `CursorModel` which has simpler named events like `next`, `first`, `last`, `previous`. How do we want to name these types of events?

Reviewer 2: Perhaps we want to align `Pagination` with `Cursor` since the concept is the same - a list of things has a cursor.

Author: Yeah, I went with `jump` and `step` as they correspond to the button names. I'm okay with switching to `first` `last`, but I'm not sure what you would name `goTo` in that paradigm.

Author: Updated. Now using `first`, `previous`, `next`, `last`, and `goTo`.

## Usage

`Pagination` is a composable component, meaning it can be put together and used in a variety of
different ways. We'll first cover basic usage examples followed by advanced usage. Each of the
examples is intended to be standalone, so you can skip to the section that best fits your use case.

These examples are designed to provide copy-and-paste snippets to help you get up and running
quickly without much explanation. Further detail on individual component props, behavior, and
accessibility can be found in the [Components](#components) section.

---

### Basic Usage

Below are examples we expect will work for most use cases. Please refer to the Advanced Usage
section for details on supporting custom implementations.

---

#### Pagination with Step Controls

In this example, the component uses "step" controls (`Pagination.StepToPreviousButton` and
`Pagination.StepToNextButton`) that allow you to move to the next page and previous pages.

```tsx
import * as React from 'react';
import {
  Pagination,
  getLastPage,
  getVisibleResultsMax,
  getVisibleResultsMin,

> } from '../lib/Pagination';

Reviewer: I think this was a copy-and-paste error. The imports in the examples should be from `@workday/canvas-kit-labs-react-pagination`.

Author: Yep. Good catch. 👍🏻

            state.currentPage,
            resultCount,
            totalCount
          )} of ${totalCount} results`
        }
      </Pagination.AdditionalDetails>
    </Pagination>
  );
};
```

---

### Advanced Usage

Below are examples for more advanced / custom usage of these components. We expect most people won't
need to read this section, but if you're needing to go beyond our basic examples or just curious,
feel free to explore this section. While these examples are not exhaustive, they provide additional
insight into what's possible.

#### Hoisted Model Pattern

If you want the `Pagination` component to handle its state and actions internally and hook into page
change events, the basic usage examples above should be sufficient. However, if you need external
access to the model, you can use the hoisted model pattern. You can create a model outside of the
component with the `usePaginationModel` hook.

For this example, we'll create a `Pagination` component that handles 128 total items with 10 items
per page. Normally this would involve some math on your part, but don't worry about that, we've got
you covered with some handy utilities

```tsx

> import {
>   getLastPage,
>   getVisibleResultsMax,
>   getVisibleResultsMin,
> } from '../lib/Pagination';

Reviewer: ```suggestion
import * as React from 'react';
import {
  Pagination,
  getLastPage,
  getVisibleResultsMax,
  getVisibleResultsMin,
  usePaginationModel,
} from '@workday/canvas-kit-labs-react-pagination';
```
It'll be a lot easier to catch these errors once these examples are actually rendered with Nicholas' docs improvements. :)

const totalCount = 128;
const resultCount = 10;
const lastPage = getLastPage(resultCount, totalCount);

const ExternalModelPagination = () => {
  // create the model

>   const model = usePaginationModel({
>     lastPage={lastPage}
>     rangeSize={3}
>     onPageChange={pageNumber => console.log(pageNumber)}
>   });

Reviewer: ```suggestion
  const model = usePaginationModel({
    lastPage,
    rangeSize: 3,
    onPageChange: pageNumber => console.log(pageNumber),
  });
```
